src/txmempool.cpp

```diff
@@ -1105,4 +1105,23 @@ void CTxMemPool::SetIsLoaded(bool loaded)
     m_is_loaded = loaded;
 }
 
+
+CTxMemPool::EpochGuard CTxMemPool::GetFreshEpoch() const
+{
+    return EpochGuard(*this);
+}
+CTxMemPool::EpochGuard::EpochGuard(const CTxMemPool& in) : pool(in)
+{
+    assert(!pool.m_has_epoch_guard);
+    ++pool.m_epoch;
+    pool.m_has_epoch_guard = true;
+}
+
+CTxMemPool::EpochGuard::~EpochGuard()
+{
+    // prevents stale results being used
+    ++pool.m_epoch;
```

Comment:
2ccb7cc

What kind of stale scenario do you envision here ? 

If taking twice the epoch is attempted, the assert will be hit and should be able to spot this kind of programming error. 

Otherwise, if you prevent `visited` returning true for past value, I think that's bettter than enforcing lock taking on epoch (because worst-case you reprocess an entry, which is better than taking a lock on the average-case, this maybe contrary to AJ opinion).

Reply:
It's not required for correctness, but I like that it enforces the invariant that if an epoch guard is not held that all mempool entries are "younger" than the mempool's m_epoch.

By "prevents stale results being used", I meant that if some nefarious caller wanted to manually check epochs outside of the context of an epoch guard, they wouldn't "gain" information about the previous traversal. Of course, this is false because they could just check if it->m_epoch == m_epoch-1, but I personally felt it was a nice way to communicate the intent. 

Also uint64_t is big, so using an extra epochs count per interval isn't going to cause any sort of issue.